### PR TITLE
Add pagination to users index view.

### PR DESCRIPTION
**Pull request by RobertCram**
PT Story: https://www.pivotaltracker.com/story/show/150853535

**Changes proposed in this pull request:**

1. Rewrite users controller `index` method to add pagination to the users index view.

2. Rewrite users `_search_form.html.haml` view to make ransack use the membership filter settings

3. Remove unnecessary scopes `are_members` and `are_not_members` on user.

Ready for review:
@patmbolger @weedySeaDragon @thesuss

### CODE REVIEW

File: app/views/users/_users_list.html.haml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#users-list
+#users_list
```
[the remark this answers is not in the document]
Comment (RobertCram) [reply app/views/users/_users_list.html.haml:1]: I changed this to be in line with earlier implementations (companies_list, membership_applications_list).

**Review by weedySeaDragon — APPROVED** [review]

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
-    @filter_ignore_membership = true
-    @filter_are_members = params.include?(:are_members) && params[:are_members] == 'true'
-    @filter_are_not_members = params.include?(:are_members) && params[:are_members] == 'false'
-
-    %w(are_members are_not_members).each do | filter |
-      if self.instance_variable_get("@filter_#{filter}")
-        @users = @users.send(filter)
-        @filter_ignore_membership = false
-      end
+    action_params, @items_count, items_per_page = process_pagination_params('company')
```
Comment (patmbolger) [inline app/controllers/users_controller.rb]: The argument to `process_pagination_params` has to be unique across all uses of that method.  Here, `user` would suffice.

(we need a story to set up a registry of entities (argument) to be used for this method to ensure uniqueness for each use.  The entity designation is used to set `session` variables so we can remember the user's `items-per-page` and search criteria (the latter needs to be recalled when changing the items-per-page).

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,22 +1,24 @@
...
     end
+    @filter_ignore_membership = !(@filter_are_members || @filter_are_not_members)
+
+    membership_filter = 'member = true' if @filter_are_members
+    membership_filter = 'member = false' if @filter_are_not_members
+
+    @q = User.ransack(action_params)
+    @users = @q.result.includes(:membership_applications).where(membership_filter).page(params[:page]).per_page(items_per_page)
 
     render partial: 'users_list', locals: { q: @q, users: @users } if request.xhr?
```
Comment (patmbolger) [inline app/controllers/users_controller.rb]: Might as well add `items_count` to the `locals` hash while we're at it.

File: app/views/users/_users_list.html.haml [hunk 4]
```diff
@@ -14,3 +14,9 @@
         %th= t('.expire_date')
     %tbody.panel-body
       = render partial: 'user', collection: users
+
+  = render partial: 'application/paginate_footer',
+           locals: { entities: @users,
+                     paginate_class: 'users_pagination',
+                     items_count: @items_count,
```
Comment (patmbolger) [inline app/views/users/_users_list.html.haml]: Use local value `items_count` here (if you are passing this in as a local var from the controller).

**Review by patmbolger — CHANGES_REQUESTED** [review]

**Comment by patmbolger** [comment]
@RobertCram - please see code comments.

In addition, column sorting (member number, email) does not work.  This is because the original implementation of users search left off the required class in the sort link in the view.

For example, in `_users_list.html.haml` changing this statement:
```
%th= sort_link(q, :padded_membership_number, t('users.users_list.membership_number'), {}, 
{ remote: true } )
```
to this:
```
%th= sort_link(q, :padded_membership_number, t('users.users_list.membership_number'), {},
          { class: 'users_pagination', remote: true } )
```
enables sorting on this column.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (RobertCram) [reply app/controllers/users_controller.rb]: A, yes, that's a copy & paste error. We do have a story somewhere to bring all the index views in line (code as well as appearance). When working on this story I realised I would like to have a helper to do most of the index view and pagination stuff for the different models. At the moment it feels like a lot of repetitive work. 

Do you think that might be an idea?

Comment (patmbolger) [reply app/controllers/users_controller.rb]: Yes, absolutely.  We've now done it enough times for a clear pattern to emerge - which we can refactor for reuse (and avoid repetition).

Comment (patmbolger) [reply app/controllers/users_controller.rb]: We do have the params processing method, and the pagination footer, which encapsulate much of the heavy lifting, but I am sure there is more opportunity to generalize further.

Comment (patmbolger) [reply app/controllers/users_controller.rb]: @RobertCram - I think you're still missing this one change (copy-and-paste error fix).  When that's done we can merge this PR.

**Review by patmbolger — APPROVED** [review]

**Pull request merged**